src/lua/uri.lua

```diff
@@ -169,8 +171,52 @@ local function format(uri, write_password)
     return str
 end
 
+local function encode_kv(key, values)
+    local val = values
+    if type(val) ~= "table" then
+        val = {val}
+    elseif table.getn(values) == 0 then
+        return ("%s"):format(key)
+    end
+    local res
+    for _, v in pairs(val) do
+        local kv = ("%s=%s"):format(key, v)
+        if not res and kv then
+            res = kv
+        elseif kv then
+            res = ("%s&%s"):format(res , kv)
+        end
+    end
+
+    return res
+end
+
+-- Encode map to a string.
+local function params(opts)
+    if not opts then
+        return ""
+    end
+    local query_string
+    for key, value in pairs(opts) do
+        if type(key) ~= "string" then
+            return nil, "Parameters keys must have a type 'string'."
+        end
+        local key_qs = encode_kv(key, value)
+        if not query_string and key_qs then
+            query_string = key_qs
+        elseif key_qs then
+            query_string = ("%s&%s"):format(query_string, key_qs)
+        end
+    end
```

Comment:
We can accumulate `'a=b'` strings into an array-like table and join them with `table.concat()`. The code will be simpler and will not generate extra intermediate strings.